crypto/hash/hmac_mbedtls.c

```diff
@@ -0,0 +1,253 @@
+
+    /* free memory */
+    srtp_crypto_free(a);
+
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_hmac_mbedtls_start(void *statev)
+{
+    mbedtls_md_context_t *state = (mbedtls_md_context_t *)statev;
+    if (mbedtls_md_hmac_reset(state) != 0)
+        return srtp_err_status_auth_fail;
+
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_hmac_mbedtls_init(void *statev,
+                                                const uint8_t *key,
+                                                int key_len)
+{
+    mbedtls_md_context_t *state = (mbedtls_md_context_t *)statev;
+    const mbedtls_md_info_t *info = NULL;
+
+    info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA1);
+    if (info == NULL)
+        return srtp_err_status_auth_fail;
+
+    if (mbedtls_md_setup(state, info, 1) != 0)
+        return srtp_err_status_auth_fail;
+
+    debug_print(srtp_mod_hmac, "mbedtls setup, name: %s",
+                mbedtls_md_get_name(info));
+    debug_print(srtp_mod_hmac, "mbedtls setup, size: %d",
+                mbedtls_md_get_size(info));
+
+    if (mbedtls_md_hmac_starts(state, key, key_len) != 0)
+        return srtp_err_status_auth_fail;
+
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_hmac_mbedtls_update(void *statev,
+                                                  const uint8_t *message,
+                                                  int msg_octets)
+{
+    mbedtls_md_context_t *state = (mbedtls_md_context_t *)statev;
+
+    debug_print(srtp_mod_hmac, "input: %s",
+                srtp_octet_string_hex_string(message, msg_octets));
+
+    if (mbedtls_md_hmac_update(state, message, msg_octets) != 0)
+        return srtp_err_status_auth_fail;
+
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_hmac_mbedtls_compute(void *statev,
+                                                   const uint8_t *message,
+                                                   int msg_octets,
+                                                   int tag_len,
+                                                   uint8_t *result)
+{
+    mbedtls_md_context_t *state = (mbedtls_md_context_t *)statev;
+    uint8_t hash_value[SHA1_DIGEST_SIZE];
+    int i;
+
+    /* check tag length, return error if we can't provide the value expected */
+    if (tag_len > SHA1_DIGEST_SIZE) {
+        return srtp_err_status_bad_param;
+    }
+
+    /* hash message, copy output into H */
+    if (mbedtls_md_hmac_update(statev, message, msg_octets) != 0)
+        return srtp_err_status_auth_fail;
+
+    if (mbedtls_md_hmac_finish(state, hash_value) != 0)
+        return srtp_err_status_auth_fail;
+
+    /* copy hash_value to *result */
+    for (i = 0; i < tag_len; i++) {
+        result[i] = hash_value[i];
+    }
+
+    debug_print(srtp_mod_hmac, "output: %s",
+                srtp_octet_string_hex_string(hash_value, tag_len));
+
+    return srtp_err_status_ok;
+}
+
+/* begin test case 0 */
```

Comment:
It seems like the test cases should be common across the various implementations.  The test data should hash to the same thing regardless of whether you're using OpenSSL, mBedTLS, etc.  Suggest factoring the tests out to a different file.  I haven't looked at the encryption code, but I suspect the same comments apply.  Perhaps we could just make a `crypto_test_cases.c` file that contains all of them.  Then the individual methods can just point to the one that makes sense.

Reply:
This idea came up when I was coding. But I saw the code inside openssl, I left it as the same as the part of openssl. Do you want to add this part, or cut another pr?

Reply:
I have a local change that does this now, I will create a PR after this is merged.